Samples/2.3/AdHocReactiveCaching/src/Grains/AggregatorCacheGrain.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Grains.Models;
+using Microsoft.Extensions.Logging;
+using Orleans;
+using Orleans.Concurrency;
+
+namespace Grains
+{
+    [Reentrant]
+    [StatelessWorker]
+    public class AggregatorCacheGrain : Grain, IAggregatorCacheGrain
+    {
+        private readonly ILogger<AggregatorCacheGrain> _logger;
+        private readonly CancellationTokenSource _token = new CancellationTokenSource();
+        private VersionedValue<int> _cache;
+        private Task _pollTask;
+
+        public AggregatorCacheGrain(ILogger<AggregatorCacheGrain> logger)
+        {
+            _logger = logger;
+        }
+
+        private string GrainType => nameof(AggregatorCacheGrain);
+        private string GrainKey => this.GetPrimaryKeyString();
+
+        public override async Task OnActivateAsync()
+        {
+            // hydrate the cache with whatever value is available right now
+            _cache = await GrainFactory.GetGrain<IAggregatorGrain>(GrainKey).GetAsync();
+
+            // start polling
+            _pollTask = PollAsync();
+
+            await base.OnActivateAsync();
+        }
+
+        public override Task OnDeactivateAsync()
+        {
+            _token.Cancel();
+
+            return base.OnDeactivateAsync();
+        }
+
+        private async Task PollAsync()
```

Comment:
Asynchronous loops like this is something that we generally don't recommend as a pattern because it is easy to get into troubles with them. The preferred way is to schedule a timer instead. Each time tick would execute an equivalent of an iteration of the loop, but with a cleaner separation of individual tasks.